contracts/src/BeefyClient.sol

```diff
@@ -323,43 +281,51 @@ contract BeefyClient {
      * @param leafProof an MMR leaf proof
      * @param leafProofOrder a bitfield describing the order of each item (left vs right)
      */
-    function submitFinalWithHandover(
+    function submitFinal(
         Commitment calldata commitment,
         uint256[] calldata bitfield,
         ValidatorProof[] calldata proofs,
         MMRLeaf calldata leaf,
         bytes32[] calldata leafProof,
         uint256 leafProofOrder
-    ) public {
+    ) external {
         (bytes32 commitmentHash, bytes32 ticketID) = validate(commitment, bitfield);
 
         Ticket storage ticket = tickets[ticketID];
 
-        if (commitment.validatorSetID != nextValidatorSet.id) {
+        if (commitment.validatorSetID != currentValidatorSet.id && commitment.validatorSetID != nextValidatorSet.id) {
             revert InvalidCommitment();
         }
-        verifyCommitment(commitmentHash, bitfield, nextValidatorSet, ticket, proofs);
 
-        if (leaf.nextAuthoritySetID != nextValidatorSet.id + 1) {
+        bool is_next_validator_set = commitment.validatorSetID == nextValidatorSet.id;
+
+        verifyCommitment(
+            commitmentHash, bitfield, is_next_validator_set ? nextValidatorSet : currentValidatorSet, ticket, proofs
+        );
+
+        if (is_next_validator_set && leaf.nextAuthoritySetID != nextValidatorSet.id + 1) {
             revert InvalidMMRLeaf();
         }
+
         bytes32 newMMRRoot = getFirstMMRRoot(commitment);
         bool leafIsValid =
             MMRProof.verifyLeafProof(newMMRRoot, keccak256(encodeMMRLeaf(leaf)), leafProof, leafProofOrder);
         if (!leafIsValid) {
             revert InvalidMMRLeafProof();
         }
 
-        currentValidatorSet = nextValidatorSet;
-        nextValidatorSet.id = leaf.nextAuthoritySetID;
-        nextValidatorSet.length = leaf.nextAuthoritySetLen;
-        nextValidatorSet.root = leaf.nextAuthoritySetRoot;
+        if (is_next_validator_set) {
```

Comment:
I think we should only verify the leaf within this block. Otherwise we're just wasting a huge amount of gas.

Reply:
As mentioned, the initial intention for this PR is  

> Also verify MMRLeaf for the non-handover case before updating MMRRoot

mainly for security considerations, gas consumed did increase but seems acceptable as follows(e.g. 3.790% for the non-handover testSubmit):

```
Test result: ok. 23 passed; 0 failed; finished in 29.97ms
testCreateInitialBitfield() (gas: -44 (-0.007%))
testCreateInitialBitfieldInvalid() (gas: -44 (-0.008%))
testRegenerateBitField() (gas: -22 (-0.018%))
testSubmitWithHandoverFailStaleCommitment() (gas: 430 (0.063%))
testScaleEncodeCommit() (gas: -22 (-0.067%))
testSubmitWithHandoverFailWithoutPrevRandao() (gas: 430 (0.077%))
testSubmitFailWithInvalidMMRLeafProof() (gas: 874 (0.116%))
testSubmitWithHandover() (gas: 857 (0.116%))
testSubmitFailWithInvalidMMRLeaf() (gas: 897 (0.120%))
testSubmitFailForPrevRandaoCapturedMoreThanOnce() (gas: 344 (0.128%))
testCreateFinalBitfield() (gas: 366 (0.131%))
testCreateFinalBitfieldInvalid() (gas: 366 (0.137%))
testSubmitWithHandoverFailWithInvalidValidatorSet() (gas: 1078 (0.161%))
testSubmitFailForPrevRandaoTooEarlyOrTooLate() (gas: 344 (0.165%))
testSubmitFailWithNotEnoughClaims() (gas: 367 (0.184%))
testSubmitFailWithStaleCommitment() (gas: 29291 (3.494%))
testSubmitFailInvalidSignature() (gas: 24694 (3.749%))
testSubmit():((uint32,uint64,(bytes2,bytes)[])) (gas: 26916 (3.790%))
testSubmitFailWithInvalidTicket() (gas: 24397 (3.813%))
testSubmitFailValidatorNotInBitfield() (gas: 24694 (3.819%))
testSubmitFailWithInvalidBitfield() (gas: 24429 (3.841%))
testSubmitFailWithInvalidValidatorSet() (gas: 24964 (3.871%))
testSubmitFailWithoutPrevRandao() (gas: 24407 (4.574%))
Overall gas change: 210013 (1.741%)
```

Reply:
It's still not clear to me why there it improves security to verify the MMR leaf in the non-handover case.  In the non-handover case, the data in the MMR leaf is not used at all, so the leaf's validity is of no importance, I'd say.

Regarding gas costs. An increase of 4% is roughly equivalent to an increase of $40,000 in yearly operating costs. So I would like to avoid extra costs if at all possible.

Reply:
> An increase of 4% is roughly equivalent to an increase of $40,000 in yearly operating costs. 

Did not realize that much! Good to know and yes we need to be super careful in gas consumption.

Reply:
I believe it's because the handover case updates `nextValidatorSet` with values from the `leaf`: https://github.com/Snowfork/snowbridge/pull/919/files#diff-473fead6cd500a4a05f06d0cea4d9ef166c9bdbe2a2e1a06a33d2e1aedcc428eR320-R325

Leaf proof verification checks that the leaf is part of a Merkle tree - it only verifies the leaf, not the `newMMRRoot`. Given that we don't use the leaf in the non-handover case, there's no need to verify it.

Reply:
That makes sense, thanks for the explanation. Revamp in https://github.com/Snowfork/snowbridge/pull/919/commits/4d7ddc3259adfde193a1a7b3b3b36737bc23c0f2

Reply:
Actually my previous thought is that in `MMRLeaf` there are fields like `parentHash` and ` parachainHeadsRoot` which will be used to verify the new `MMRRoot` even in non-handover case.   I would assume though the proof can be faked but adds some complexity for bad relayers.

If you all think that does not make much sense, then I would prefer to abandon the improvement here. Because now always require `MMRLeaf` as call data which costs much more gas(3%-4%).

Reply:
I don't think we can trust the "gas diff". because it also includes gas executed during test setup.

You should run `forge test --match-path test/BeefyClient.t.sol --gas-report`, which gives more info:
```
 test/mocks/BeefyClientMock.sol:BeefyClientMock contract |                 |       |        |        |         |
|---------------------------------------------------------|-----------------|-------|--------|--------|---------|
| Deployment Cost                                         | Deployment Size |       |        |        |         |
| 2186482                                                 | 11250           |       |        |        |         |
| Function Name                                           | min             | avg   | median | max    | # calls |
| commitPrevRandao                                        | 785             | 17943 | 21119  | 21119  | 19      |
| createFinalBitfield                                     | 1134            | 9966  | 10597  | 10597  | 15      |
| createInitialBitfield                                   | 544             | 52136 | 52804  | 70348  | 52      |
| encodeCommitment_public                                 | 5556            | 5556  | 5556   | 5556   | 1       |
| initialize_public                                       | 2281            | 84897 | 92081  | 92081  | 25      |
| latestBeefyBlock                                        | 466             | 466   | 466    | 466    | 3       |
| minimumSignatureThreshold_public                        | 267             | 267   | 267    | 267    | 1       |
| setLatestBeefyBlock                                     | 20364           | 20364 | 20364  | 20364  | 1       |
| submitFinal                                             | 4569            | 31392 | 6444   | 108887 | 9       |
| submitFinalWithHandover                                 | 5071            | 52439 | 48615  | 112192 | 6       |
| submitInitial                                           | 5625            | 53973 | 61254  | 61254  | 15      |
| submitInitialWithHandover                               | 5515            | 53197 | 61144  | 61144  | 7       |
```

Reply:
I have an idea. According to the Ethereum spec:

> The Ethereum Virtual Machine imposes a cost of 4 gas for every byte set to zero in calldata, and 16 gas for every non-zero byte (reference: [Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf)).

If for the non-handover case, the relayer passes **zero values** for `leaf` and `leafProof` when calling `submitFinal`, then the increase in gast cost would be minimal.

Can you benchmark this approach to find out?

Reply:
|---------------------------------------------------------|-----------------|-------|--------|--------|---------|
| Deployment Cost                                         | Deployment Size |       |        |        |         |
| 2197297                                                 | 11304           |       |        |        |         |
| Function Name                                           | min             | avg   | median | max    | # calls |
| commitPrevRandao                                        | 763             | 17921 | 21097  | 21097  | 19      |
| createFinalBitfield                                     | 1134            | 9966  | 10597  | 10597  | 15      |
| createInitialBitfield                                   | 544             | 52080 | 52804  | 70348  | 48      |
| encodeCommitment_public                                 | 5556            | 5556  | 5556   | 5556   | 1       |
| initialize_public                                       | 2259            | 84250 | 92059  | 92059  | 23      |
| latestBeefyBlock                                        | 488             | 488   | 488    | 488    | 3       |
| minimumSignatureThreshold_public                        | 267             | 267   | 267    | 267    | 1       |
| setLatestBeefyBlock                                     | 20364           | 20364 | 20364  | 20364  | 1       |
| submitFinal                                             | 4829            | 40169 | 6704   | 112156 | 15      |
| submitInitial                                           | 14308           | 58987 | 61586  | 61853  | 20      |